### PR TITLE
Adapt KDE Plasma5 for CheriABI

### PR DESCRIPTION
**Pull request by bukinr**

### CODE REVIEW

File: devel/kf5-ki18n/Makefile [hunk 1]
```diff
@@ -8,7 +8,13 @@ COMMENT=	KF5 advanced internationalization framework
...
 		gettext-tools:build,run kde:5 qt:5 tar:xz
 USE_KDE=	ecm:build
-USE_QT=		concurrent core declarative network script \
+USE_QT=		concurrent core declarative network \
 		buildtools:build qmake:build testlib:build
 
-.include <bsd.port.mk>
+.include <bsd.port.pre.mk>
+
+.if ${ARCH} != aarch64c && ${ARCH} != riscv64c
```
Comment (kwitaszczyk) [inline devel/kf5-ki18n/Makefile:16]: Here and in other places:
```
.if !${ABI:Mpurecap}
```

File: devel/kf5-ki18n/Makefile [hunk 2]
```diff
@@ -8,7 +8,13 @@ COMMENT=	KF5 advanced internationalization framework
...
+USES+=	script
```
Comment (kwitaszczyk) [inline devel/kf5-ki18n/Makefile:17]:
```
USE_QT+=    script
```

File: devel/kf5-kio/Makefile [hunk 3]
```diff
@@ -6,18 +6,37 @@ CATEGORIES=	devel kde kde-frameworks
...
 USE_XORG=	x11
 
-.include <bsd.port.mk>
+OPTIONS_DEFINE=		DOCBOOK
+OPTIONS_DEFAULT=	DOCBOOK
+OPTIONS_SUB=		yes
+DOCBOOK_DESC=		Doctools build
+
+OPTIONS_EXCLUDE_aarch64c=	DOCBOOK
+OPTIONS_EXCLUDE_riscv64c=	DOCBOOK
```
Comment (kwitaszczyk) [inline devel/kf5-kio/Makefile]: Here and in other places: use `OPTIONS_EXCLUDE_purecap`.

File: devel/kf5-kio/Makefile [hunk 4]
```diff
@@ -6,18 +6,37 @@ CATEGORIES=	devel kde kde-frameworks
...
-		doctools:build ecm:build
+		ecm:build
 USE_QT=		concurrent core dbus declarative gui network \
-		script widgets x11extras xml \
+		widgets x11extras xml \
 		buildtools:build qmake:build testlib:build
 USE_XORG=	x11
 
-.include <bsd.port.mk>
+OPTIONS_DEFINE=		DOCBOOK
```
Comment (kwitaszczyk) [inline devel/kf5-kio/Makefile]: Here and in other places: `DOCBOOK` -> `DOCS`?

File: Mk/Uses/cmake.mk [hunk 5]
```diff
@@ -44,6 +44,9 @@
 #
 # MAINTAINER: kde@FreeBSD.org
 
+# CHERI: c64 msgfmt is broken, use a64 version globally
```
Comment (kwitaszczyk) [inline Mk/Uses/cmake.mk:47]:
```
# CheriABI msgfmt is broken. Use a hybrid ABI version instead globally.
```

File: Mk/Uses/cmake.mk [hunk 6]
```diff
@@ -44,6 +44,9 @@
...
+CMAKE_ARGS+=	-DGETTEXT_MSGFMT_EXECUTABLE=/usr/local64/bin/msgfmt
```
Comment (kwitaszczyk) [inline Mk/Uses/cmake.mk:48]: This should be moved after
```
.f !defined(_INCLUDE_USES_CMAKE_MK)
 _INCLUDE_USES_CMAKE_MK=	yes
```
and even better next to `CMAKE_ARGS`, only for a CheriABI build and using a predefined LOCALBASE64 instead of hardcoding a path:
```
.if ${ARCH:Mpurecap}
CMAKE_ARGS+=    -DGETTEXT_MSGFMT_EXECUTABLE=${LOCALBASE64}/bin/msgfmt
.endif
```
However, `msgfmt` is not always required. It should be required when `gettext-tools` is added to `USES`. It would be better to move this logic to `Mk/Uses/gettext-tools.mk` with a condition that `cmake` is also used, i.e.:
```
.if ${USES:Mcmake} && ${ARCH:Mpurecap}
CMAKE_ARGS+=    -DGETTEXT_MSGFMT_EXECUTABLE=${LOCALBASE64}/bin/msgfmt
.endif
```

This logic assumes that you have a hybrid ABI msgfmt installed which is not true when building packages and possibly when a user wants to build ports manually. In order to use a hybrid ABI msgfmt, we must mark `devel/gettext-tools` as broken for purecap and allow to use a replacement hybrid ABI package instead, i.e. in `devel/gettext-tools/Makefile`, add:
```
BROKEN_purecap=   msgfmt is broken for CheriABI
USE_PKG64=        1
```

File: Mk/Uses/cmake.mk (hunk 5 above)
Comment (kwitaszczyk) [inline Mk/Uses/cmake.mk:47]: `Mk/Uses/cmake.mk` in the commit message.

File: x11-wm/plasma5-kwin/files/cheribsd.patch [hunk 7]
```diff
@@ -0,0 +1,201 @@
+diff -ru CMakeLists.txt CMakeLists.txt
```
Comment (kwitaszczyk) [inline x11-wm/plasma5-kwin/files/cheribsd.patch:1]: Did you generate this patch from some branch? If so, it would be good to reference it in the commit message.

File: x11-wm/plasma5-kwin/Makefile [hunk 8]
```diff
@@ -41,14 +39,35 @@ USE_KDE=	activities attica auth breeze codecs completion config \
...
+OPTIONS_EXCLUDE_aarch64c=	DOCBOOK
+OPTIONS_EXCLUDE_riscv64c=	DOCBOOK
+
+.include <bsd.port.pre.mk>
+
+.if ${ARCH} != aarch64c && ${ARCH} != riscv64c
+BUILD_DEPENDS+=	xwayland-devel>0:x11-servers/xwayland-devel
+RUN_DEPENDS+=	xwayland-devel>0:x11-servers/xwayland-devel
+USES+=		python:3.7+,run gnome
+USE_QT+=	multimedia
```
Comment (kwitaszczyk) [inline x11-wm/plasma5-kwin/Makefile:62]:
```
USE_QT+=	multimedia script
```

File: x11-wm/plasma5-kwin/Makefile [hunk 9]
```diff
@@ -41,14 +39,35 @@ USE_KDE=	activities attica auth breeze codecs completion config \
...
+
+OPTIONS_EXCLUDE_aarch64c=	DOCBOOK
+OPTIONS_EXCLUDE_riscv64c=	DOCBOOK
+
+.include <bsd.port.pre.mk>
+
+.if ${ARCH} != aarch64c && ${ARCH} != riscv64c
+BUILD_DEPENDS+=	xwayland-devel>0:x11-servers/xwayland-devel
+RUN_DEPENDS+=	xwayland-devel>0:x11-servers/xwayland-devel
+USES+=		python:3.7+,run gnome
```
Comment (kwitaszczyk) [inline x11-wm/plasma5-kwin/Makefile:61]:
```
USES+=		gnome python:3.7+,run
```

File: devel/kf5-ktexteditor/Makefile [hunk 10]
```diff
@@ -17,8 +14,24 @@ USE_KDE=	archive auth codecs completion config configwidgets \
...
 USE_XORG=	x11
 
-.include <bsd.port.mk>
+OPTIONS_DEFINE=		POLKIT
+OPTIONS_DEFAULT=	POLKIT
+OPTIONS_SUB=		yes
+DOCBOOK_DESC=		Polkit Qt support
+
+OPTIONS_EXCLUDE_aarch64c=	POLKIT
+OPTIONS_EXCLUDE_riscv64c=	POLKIT
```
Comment (kwitaszczyk) [inline devel/kf5-ktexteditor/Makefile:27]: Is Polkit fundamentally broken for CheriABI? If so, what do you think about disabling it globally for purecap?

File: security/plasma5-kscreenlocker/Makefile [hunk 11]
```diff
@@ -29,4 +27,19 @@ CPE_VENDOR=	kde
...
 SHEBANG_FILES=	updaters/ksreenlocker_5_3_separate_autologin.pl
 
-.include <bsd.port.mk>
+OPTIONS_DEFINE=		CONSOLEKIT
+OPTIONS_DEFAULT=	CONSOLEKIT
+OPTIONS_SUB=		yes
+DOCBOOK_DESC=		Depend on Console Kit
+
+OPTIONS_EXCLUDE_aarch64c=	CONSOLEKIT
+OPTIONS_EXCLUDE_riscv64c=	CONSOLEKIT
```
Comment (kwitaszczyk) [inline security/plasma5-kscreenlocker/Makefile:36]: Is Console Kit fundamentally broken for CheriABI? If so, what do you think about disabling it globally for purecap?

File: devel/kf5-kio/Makefile [hunk 12]
```diff
@@ -6,18 +6,37 @@ CATEGORIES=	devel kde kde-frameworks
...
+OPTIONS_SUB=		yes
+DOCBOOK_DESC=		Doctools build
+
+OPTIONS_EXCLUDE_aarch64c=	DOCBOOK
+OPTIONS_EXCLUDE_riscv64c=	DOCBOOK
+
+.include <bsd.port.pre.mk>
+
+.if ${ARCH} != aarch64c && ${ARCH} != riscv64c
+USES+=		desktop-file-utils gnome
```
Comment (kwitaszczyk) [inline devel/kf5-kio/Makefile:30]: I wonder in what cases an excluded USE should become a disabled option. Do you have any thoughts on that?

File: devel/libudev-devd/files/cheribsd.patch [hunk 13]
```diff
@@ -0,0 +1,13 @@
+diff --git udev-device.c udev-device.c
+index 9f163d1..3ad05d2 100644
+--- udev-device.c
++++ udev-device.c
+@@ -180,6 +180,8 @@ udev_device_has_tag(struct udev_device *ud, const char *tag)
+ {
+ 	struct udev_list_entry *ule;
+ 
++	return (0);
++
```
Comment (kwitaszczyk) [inline devel/libudev-devd/files/cheribsd.patch:10]: There should be a comment explaining this change, and possibly a better solution that removes the whole function contents, including an unused `ule` variable.

File: devel/qt5/files/extrapatch-cheribsd.patch [hunk 14]
```diff
@@ -1,3 +1,23 @@
+diff --git bin/fixqt4headers.pl bin/fixqt4headers.pl
+index b572f469dc..777dff88d1 100755
+--- bin/fixqt4headers.pl
++++ bin/fixqt4headers.pl
+@@ -1,4 +1,4 @@
+-#!/usr/bin/env perl
++#!/usr/local64/bin/perl
```
Comment (kwitaszczyk) [inline devel/qt5/files/extrapatch-cheribsd.patch:7]: Have you verified if it breaks with the current CheriABI Perl that had several CheriABI-related issues solved?

File: graphics/libepoxy/Makefile [hunk 15]
```diff
@@ -15,8 +15,8 @@ USE_GL=		glesv2 egl
 USE_LDCONFIG=	yes
 SHEBANG_FILES=	src/gen_dispatch.py
 
-OPTIONS_DEFINE=		TEST X11
-OPTIONS_DEFAULT=	X11
+OPTIONS_DEFINE=		TEST X11 EGL
+OPTIONS_DEFAULT=	X11 EGL
```
Comment (kwitaszczyk) [inline graphics/libepoxy/Makefile:19]: We should alphabetically sort such lists.

Comment (kwitaszczyk) [inline graphics/libepoxy/Makefile:19]: Should `EGL` be enabled for `aarch64` and `aarch64c` or all architectures? We should consider what CheriBSD-specific changes should be upstreamed.

File: textproc/kf5-syntax-highlighting/Makefile [hunk 16]
```diff
@@ -12,4 +12,10 @@ USE_QT=		core gui network \
 
 CMAKE_ARGS=	-DCMAKE_DISABLE_FIND_PACKAGE_Qt5XmlPatterns=TRUE
 
-.include <bsd.port.mk>
+.include <bsd.port.pre.mk>
+
+.if ${ARCH} == aarch64c || ${ARCH} == riscv64c
+CMAKE_ARGS+=	-DPERL_EXECUTABLE=/usr/local64/bin/perl
```
Comment (kwitaszczyk) [inline textproc/kf5-syntax-highlighting/Makefile:18]: What bug did you hit here with CheriABI Perl?

At the moment, there's no dependency on hybrid ABI Perl here. It means that a CheriABI build of this port would fail due to missing hybrid ABI Perl. I wouldn't want to use the same trick with Perl as we do with Python as many ports work fine with CheriABI Perl. We should try to fix the bug that occurs here.

File: devel/kf5-kio/Makefile [hunk 17]
```diff
@@ -20,13 +20,9 @@ USE_QT=		concurrent core dbus declarative gui network \
 		buildtools:build qmake:build testlib:build
 USE_XORG=	x11
 
-OPTIONS_DEFINE=		DOCBOOK
-OPTIONS_DEFAULT=	DOCBOOK
+OPTIONS_DEFINE=		DOCS
```
Comment (kwitaszczyk) [inline devel/kf5-kio/Makefile:23]: Please fix-up this change and other "`DOCBOOK` -> `DOCS`" changes with previous commits.

File: graphics/mesa-dri/Makefile [hunk 18]
```diff
@@ -72,10 +74,20 @@ USE_XORG+=		xcb xrandr
...
+		-Dplatforms="${MESA_PLATFORMS:ts,:tl}" \
+		-Degl=disabled
+.else
+MESON_ARGS+=	-Ddri-drivers="" \
+		-Dgallium-drivers="${GALLIUM_DRIVERS:ts,:tl}" \
+		-Dvulkan-drivers="" \
+		-Dplatforms="${MESA_PLATFORMS:ts,:tl}" \
+		-Dllvm=disabled \
+		-Degl=enabled
+.endif
```
Comment (kwitaszczyk) [inline graphics/mesa-dri/Makefile:90]:
```
.if ${ABI:Mpurecap}
DRI_DRIVERS=""
VULKAN_DRIVERS=""
.endif

(...)

MESON_ARGS+=	-Ddri-drivers="${DRI_DRIVERS:ts,:tl}" \
 		-Dgallium-drivers="${GALLIUM_DRIVERS:ts,:tl}" \
		-Dvulkan-drivers="${VULKAN_DRIVERS:ts,:tl}" \
		-Dplatforms="${MESA_PLATFORMS:ts,:tl}"
.if ${ABI:Mpurecap}
MESON_ARGS+=	-Degl=enabled
.else
MESON_ARGS+=	-Degl=disabled \
		-Dllvm=disabled
.endif
```

File: graphics/mesa-libs/Makefile [hunk 19]
```diff
@@ -62,6 +60,15 @@ MESA_PLATFORMS+=	wayland
 
 MESON_ARGS+=	-Dplatforms="${MESA_PLATFORMS:ts,:tl}"
 
+.if ${ARCH} == aarch64c || ${ARCH} == riscv64c
+MESON_ARGS+=	-Dllvm=disabled -Dgles1=enabled -Dgles2=enabled
+MESON_ARGS+=   -Ddri-drivers=""
+MESON_ARGS+=   -Dgallium-drivers=panfrost,swrast
```
Comment (kwitaszczyk) [inline graphics/mesa-libs/Makefile:66]: It would be better to set `DRI_DRIVERS` and `GALLIUM_DRIVERS` before `MESON_ARGS`, as I suggested in `graphics/mesa-dri`.

File: x11-fonts/fontconfig/files/cheribsd.patch [hunk 20]
```diff
@@ -395,3 +395,62 @@ index 82c6bed..e1cd507 100644
...
+     x ^= x >> 27;
+@@ -117,7 +117,7 @@
+ static FcSerializeBucket*
+ FcSerializeFind (const FcSerialize *serialize, const void *object)
+ {
+-    uintptr_t hash = FcSerializeHashPtr (object);
++    ptraddr_t hash = FcSerializeHashPtr (object);
+     size_t buckets_count = serialize->buckets_count;
+     size_t index = hash & (buckets_count-1);
+     for (size_t n = 0; n < buckets_count; ++n) {
```
Comment (kwitaszczyk) [inline x11-fonts/fontconfig/files/cheribsd.patch:456]: Add these changes to https://github.com/CTSRD-CHERI/fontconfig/tree/2.14.0-cheriabi and regenerate the patch, please.

File: www/firefox/files/cheribsd.patch [hunk 21]
```diff
@@ -0,0 +1,18 @@
+--- js/src/gc/Memory.cpp	2022-09-26 12:03:35.821615000 +0100
++++ js/src/gc/Memory.cpp	2022-09-26 12:03:57.178195000 +0100
+@@ -442,6 +442,7 @@
+   return region;
+ #else
+ 
++#if 0
```
Comment (kwitaszczyk) [inline www/firefox/files/cheribsd.patch:7]: It would be good to add a comment why the scattershot allocator must be disabled so that we know what should be fixed in Firefox/CheriBSD.

File: devel/kf5-extra-cmake-modules/Makefile [hunk 22]
```diff
@@ -21,6 +21,7 @@ PORTDOCS=	html/*
 OPTIONS_DEFINE=	DOCS HELP MANPAGES
 OPTIONS_DEFAULT=HELP MANPAGES
 OPTIONS_SUB=	yes
+OPTIONS_EXCLUDE_purecap=DOCS HELP MANPAGES
```
Comment (kwitaszczyk) [inline devel/kf5-extra-cmake-modules/Makefile]: You can assume that `DOCS` is disabled:
https://github.com/CTSRD-CHERI/poudriere/blob/master/src/etc/poudriere.d/cheriabi-make.conf.sample

File: sysutils/plasma5-libksysguard/Makefile [hunk 23]
```diff
@@ -23,4 +23,17 @@ WEBENGINE_USE=	QT=webengine
...
+OPTIONS_DEFAULT=	POLKIT
+OPTIONS_SUB=		yes
+DOCBOOK_DESC=		Polkit Qt support
+
+OPTIONS_EXCLUDE_purecap=      POLKIT
+
+.include <bsd.port.options.mk>
+
+.if ${ARCH} != aarch64c && ${ARCH} != riscv64c
+USES+=	script
```
Comment (kwitaszczyk) [inline sysutils/plasma5-libksysguard/Makefile:36]:
```
USES+=   script webkit
```

File: sysutils/plasma5-kinfocenter/files/cheribsd.patch [hunk 24]
```diff
@@ -0,0 +1,18 @@
+--- CMakeLists.txt	2022-09-30 17:28:28.943923000 +0100
```
Comment (kwitaszczyk) [inline sysutils/plasma5-kinfocenter/files/cheribsd.patch:1]: As we talked during the last CheriBSD packages call, I was thinking about a better solution for this. At the moment, this port wouldn't build for aarch64 when DOCS is enabled as plist entries would be missing.

Instead, we could rename this patch to `cheribsd-docs.patch` and apply it if DOCS is enabled. Based on `devel/radare2/Makefile`, we could do:
```
DOCS_EXTRA_PATCHES_OFF= ${PATCHDIR}/cheribsd-docs.patch
```

File: sysutils/plasma5-kde-cli-tools/Makefile [hunk 25]
```diff
@@ -8,18 +8,31 @@ COMMENT=	Plasma5 non-interactive system tools
 WWW=		https://www.kde.org/plasma-desktop
 
 USES=		cmake compiler:c++11-lib cpe gettext desktop-file-utils \
-		 kde:5 qt:5 tar:xz xorg
+		kde:5 qt:5 tar:xz xorg
```
Comment (kwitaszczyk) [inline sysutils/plasma5-kde-cli-tools/Makefile:11]: Unnecessary change.

File: sysutils/plasma5-libksysguard/Makefile [hunk 26]
```diff
@@ -26,7 +26,7 @@ WEBENGINE_CMAKE_BOOL_OFF=	CMAKE_DISABLE_FIND_PACKAGE_Qt5WebEngineWidgets
 OPTIONS_DEFINE=		POLKIT
 OPTIONS_DEFAULT=	POLKIT
 OPTIONS_SUB=		yes
-DOCBOOK_DESC=		Polkit Qt support
+POLKIT_DESC=		Polkit Qt support
```
Comment (kwitaszczyk) [inline sysutils/plasma5-libksysguard/Makefile:29]: The commit message is invalid. Instead, the change DOCBOOK->POLKIT should be reverted in the previous commit with fix-up.

File: security/plasma5-kscreenlocker/Makefile [hunk 27]
```diff
@@ -30,7 +30,7 @@ SHEBANG_FILES=	updaters/ksreenlocker_5_3_separate_autologin.pl
 OPTIONS_DEFINE=		CONSOLEKIT
 OPTIONS_DEFAULT=	CONSOLEKIT
 OPTIONS_SUB=		yes
-DOCBOOK_DESC=		Depend on Console Kit
+CONSOLEKIT_DESC		Depend on Console Kit
```
Comment (kwitaszczyk) [inline security/plasma5-kscreenlocker/Makefile:33]: Fix-up this change, please.

File: x11/morello-desktop/Makefile [hunk 28]
```diff
@@ -0,0 +1,28 @@
+PORTNAME=	morello-desktop
```
Comment (kwitaszczyk) [inline x11/morello-desktop/Makefile:1]: This change should be the last in the branch.

File: devel/kf5-extra-cmake-modules/Makefile (hunk 22 above)
Comment (bukinr) [reply devel/kf5-extra-cmake-modules/Makefile]: Thanks. This already resolved.

File: sysutils/plasma5-libksysguard/Makefile (hunk 23 above)
Comment (bukinr) [reply sysutils/plasma5-libksysguard/Makefile]: Thanks. I missed that

File: sysutils/plasma5-kinfocenter/files/cheribsd.patch (hunk 24 above)
Comment (bukinr) [reply sysutils/plasma5-kinfocenter/files/cheribsd.patch]: Nice hint. Thanks!

File: x11/morello-desktop/Makefile (hunk 28 above)
[the remark this answers is not in the document]
Comment (bukinr) [reply x11/morello-desktop/Makefile:1]: easy but I only worry that it includes panfrost userspace drivers that are not needed in case of VNC ?

File: x11-fonts/fontconfig/files/cheribsd.patch (hunk 20 above)
Comment (bukinr) [reply x11-fonts/fontconfig/files/cheribsd.patch]: Done. Thanks!

File: graphics/libepoxy/Makefile (hunk 15 above)
Comment (bukinr) [reply graphics/libepoxy/Makefile]: I don't know exactly. Lets leave it for aarch64* only ?

File: textproc/kf5-syntax-highlighting/Makefile (hunk 16 above)
Comment (bukinr) [reply textproc/kf5-syntax-highlighting/Makefile:18]: I can confirm is still breaks during PHP code generation.

File: www/firefox/files/cheribsd.patch (hunk 21 above)
Comment (bukinr) [reply www/firefox/files/cheribsd.patch:7]: ok I found that bug on freebsd/aarch64, not cheribsd. Firefox crashes early on start sometimes. Disabling scattershot helps.

File: devel/libudev-devd/files/cheribsd.patch (hunk 13 above)
Comment (bukinr) [reply devel/libudev-devd/files/cheribsd.patch:10]: kwin does not start without this hack. I think it could not find/initialize input devices

File: www/firefox/files/cheribsd.patch (hunk 21 above)
Comment (bukinr) [reply www/firefox/files/cheribsd.patch:12]: I added a comment. Investigations are needed but that is not specifically cheribsd issue. I found this on freebsd.

File: x11/morello-desktop/Makefile [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+PORTVERSION=	20221005
+PORTREVISION=	1
+CATEGORIES=	x11
+
+MAINTAINER=	br@FreeBSD.org
+COMMENT=	Metaport of Morello Pure-Capability desktop, Wayland only.
```
[the remark this answers is not in the document]
Comment (kwitaszczyk) [reply x11/morello-desktop/Makefile]: The comment still must be updated.

File: x11/morello-desktop/Makefile [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+WWW=		https://github.com/CTSRD-CHERI/cheribsd
+
```
[the remark this answers is not in the document]
Comment (kwitaszczyk) [reply x11/morello-desktop/Makefile]: This change was not applied.

File: devel/qt5/files/extrapatch-cheribsd.patch (hunk 14 above)
Comment (kwitaszczyk) [reply devel/qt5/files/extrapatch-cheribsd.patch:7]: OK, let's debug it separately after the merge.
